### PR TITLE
Formatting and Saving Forecast Data with Tests

### LINKED ISSUE #4: format into Forecast object
After pulling the NESO forecast, it would be good to format into OCF format. 
It would be a [Forecast](https://github.com/openclimatefix/nowcasting_datamodel/blob/main/nowcasting_datamodel/models/forecast.py#L481) object

### LINKED ISSUE #5: save
Svaing the forecast can be done using https://github.com/openclimatefix/nowcasting_datamodel/blob/main/nowcasting_datamodel/save/save.py#L21

### PR DESCRIPTION
**Pull request by siddharth7113**
# Pull Request

## Description

This pull request addresses the functionality required for:
- Formatting NESO solar forecast data into OCF-compatible Forecast objects.
- Saving formatted forecasts to the database.

It tackles the following issues:
- Fixes #4
- Fixes #5

Key changes include:
- Added `format_to_forecast_sql` to transform fetched forecast data into `ForecastSQL` objects.
- Implemented `save_forecasts_to_db` to persist these formatted forecasts in the database.

- Applied code formatting (Black) and addressed all linting issues (Ruff).

## How Has This Been Tested?

Tests were added for both:
1. Formatting function (`format_to_forecast_sql`).
2. Saving function (`save_forecasts_to_db`).

The tests verify:
- The correctness of data transformation and structure.
- The ability to save real forecast data into the database.

To reproduce:
1. Run `pytest tests/test_format_forecast.py` to validate the formatting logic.
2. Run `pytest tests/test_save_forecasts.py` to validate the saving logic.

## Checklist:

- [x] My code follows [[OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)](https://github.com/openclimatefix/.github/blob/main/coding_style.md).
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have checked my code and corrected any misspellings.

### CODE REVIEW

**Comment by siddharth7113** [comment]
Hi @peterdudfield ,

I wanted to let you know that I have tested the functionality locally by running the following commands:
- `pytest tests/test_format_forecast.py` for the formatting logic.
- `pytest tests/test_save_forecasts.py` for saving the forecasts to the database.

Both tests passed successfully, and I verified the database entries for accuracy. However, as this is my first time writing tests, I approached this with a lot of learning. I referenced the project's documentation, general testing guidelines, and external resources (like GPT) to ensure I followed best practices. 

While I’ve made my best effort, I’m open to any feedback or suggestions for improvement. Please let me know if there are specific areas I should revisit or refine—I’d be happy to make the necessary adjustments.

Thank you for your guidance and support!

**Comment by siddharth7113** [comment]
Hi @peterdudfield,
I wanted to clarify the commit history. During the development of this feature, I encountered issues with the CI configuration (`pytest.yaml`). I initially tried to fix it but made an invalid configuration, which required a revert. This created extra commits in the history.

I've resolved the issue now, and the CI pipeline should function as expected. Let me know if you'd like me to squash these commits to clean up the history.

**Comment by peterdudfield** [comment]
Hey this looks really great. Thanks so much fir doing thus. Some quick points

Could you add some requirements to tge pyprojrct ti get the ci working

In the format stage, I think we should just make 1 forecasts object, which had many Forecast Values. Sorry if it does that already

**Comment by siddharth7113** [comment]
Hey @peterdudfield, thanks for the feedback and suggestions! I’m currently out of town for a few days and don’t have access to everything I need to make the updates. I’ll address this and update the PR in the next few days. Thanks for your patience!

**Comment by siddharth7113** [comment]
Hi @peterdudfield 

Thank you for the feedback! I’ve been working on ensuring that the formatting stage creates a single ForecastSQL object containing multiple ForecastValue entries.

Currently, I’m encountering an issue where multiple ForecastSQL objects are being created instead of just one. The problem seems to stem from the get_location function, which sometimes results in duplicate LocationSQL entries being added to the session, even though the gsp_id is the same. Since get_location is part of the library, I’m unable to modify it directly.

Steps Taken So Far:
I ensured get_location is called only once for gsp_id=0 before creating the ForecastSQL object.
I checked that the forecast_values list is correctly populated with all the rows.
Despite these changes, when I add the ForecastSQL object to the session, I see multiple entries.
Would you have any advice or suggestions for:

Avoiding duplicate ForecastSQL objects when using get_location?
Ensuring that ForecastSQL interacts cleanly with the session to maintain uniqueness?
Any help would be greatly appreciated!

**Comment by siddharth7113** [comment]
> > Hi @peterdudfield
> > I’ve been working on ensuring that the formatting stage creates a single`ForecastSQL` object containing multiple `ForecastValue `entries.
> > Currently, I’m encountering an issue where multiple `ForecastSQL` objects are being created instead of just one. The problem seems to stem from the `get_location` function, which sometimes results in duplicate `LocationSQL` entries being added to the session, even though the `gsp_id` is the same. Since `get_location` is part of the library, I’m unable to modify it directly.
> > Steps Taken So Far:
> > I ensured `get_location` is called only once for gsp_id=0 before creating the `ForecastSQL` object.
> > I checked that the `forecast_values` list is correctly populated with all the rows.
> > Despite these changes, when I add the `ForecastSQL` object to the session, I see multiple entries.
> > Would you have any advice or suggestions for:
> > Avoiding duplicate `ForecastSQL` objects when using `get_location`? Ensuring that `ForecastSQL` interacts cleanly with the session to maintain uniqueness?
> > Any help would be greatly appreciated!
> 
> Thats not a problem, just make one. The `save` method I think duplicated the one, so you might end up with two. Feel free to push code, and I can try to help


I've updated both the `format_forecast` code and the corresponding test code.

However, the test code was initially written to test locally with a hardcoded database configuration (e.g., localhost). It’s not well-suited for CI as it currently relies on a local PostgreSQL setup. I’ve been unable to figure out how to rewrite it for CI yet.

Would it be okay to push the current changes as they are, and we can address the test improvements (to make them CI-ready) in a separate PR?

Let me know what you think!

File: neso_solar_consumer/format_forecast.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        input_data_last_updated=input_data_last_updated,
+        forecast_values=forecast_values,
+        historic=False,
+    )
+    logger.debug(f"ForecastSQL Object Created: {forecast}")
+
+    # Step 6: Add to session and flush
+    session.add(forecast)
+    session.flush()
+    logger.info("ForecastSQL object successfully added to session and flushed.")
```
Comment (peterdudfield) [inline neso_solar_consumer/format_forecast.py:73]: can you remove this bit, the save part will save it to the database

File: tests/test_format_forecast.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    Database connection details (modify if required):
+        - Host: localhost
+        - User: postgres
+        - Password: 12345
+        - Database: testdb
+
+    Creates fresh tables before each test and cleans up after execution.
+    """
+    engine = create_engine("postgresql://postgres:12345@localhost/testdb")
```
Comment (peterdudfield) [inline tests/test_format_forecast.py:63]: this looks good, you can do something like this ` with PostgresContainer("postgres:15.5") as postgres:` that just spins it up automatically

see https://github.com/openclimatefix/pv-site-datamodel/blob/main/tests/conftest.py#L26C4-L26C57

Btw, this is hard, so dont worry if this takes some time

**Comment by peterdudfield** [comment]
> > > Hi @peterdudfield
> > > I’ve been working on ensuring that the formatting stage creates a single`ForecastSQL` object containing multiple `ForecastValue `entries.
> > > Currently, I’m encountering an issue where multiple `ForecastSQL` objects are being created instead of just one. The problem seems to stem from the `get_location` function, which sometimes results in duplicate `LocationSQL` entries being added to the session, even though the `gsp_id` is the same. Since `get_location` is part of the library, I’m unable to modify it directly.
> > > Steps Taken So Far:
> > > I ensured `get_location` is called only once for gsp_id=0 before creating the `ForecastSQL` object.
> > > I checked that the `forecast_values` list is correctly populated with all the rows.
> > > Despite these changes, when I add the `ForecastSQL` object to the session, I see multiple entries.
> > > Would you have any advice or suggestions for:
> > > Avoiding duplicate `ForecastSQL` objects when using `get_location`? Ensuring that `ForecastSQL` interacts cleanly with the session to maintain uniqueness?
> > > Any help would be greatly appreciated!
> > 
> > 
> > Thats not a problem, just make one. The `save` method I think duplicated the one, so you might end up with two. Feel free to push code, and I can try to help
> 
> I've updated both the `format_forecast` code and the corresponding test code.
> 
> However, the test code was initially written to test locally with a hardcoded database configuration (e.g., localhost). It’s not well-suited for CI as it currently relies on a local PostgreSQL setup. I’ve been unable to figure out how to rewrite it for CI yet.
> 
> Would it be okay to push the current changes as they are, and we can address the test improvements (to make them CI-ready) in a separate PR?
> 
> Let me know what you think!

Thanks so much for this, ive added a comment or two, that should actually get it working in CI. I reckon lets try and see if we can do it in this PR

File: neso_solar_consumer/format_forecast.py (hunk 1 above)
Comment (siddharth7113) [reply neso_solar_consumer/format_forecast.py:73]: Sure ,I would make the necessary changes ; Before pushing I was testing things, I forgot to remove this.

File: tests/test_fetch_data.py [hunk 3]
```diff
@@ -1,74 +1,44 @@
-"""
```
Comment (peterdudfield) [inline tests/test_fetch_data.py:1]: was there a reason to remove this text?

Comment (siddharth7113) [reply tests/test_fetch_data.py:1]: Apologies, this was unintentional. I mistakenly removed the documentation while adjusting the file for the `conftest `setup and cleaning up sections to align with it. I'll restore the docstring right away.

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    This fixture:
+    - Creates a fresh database schema before each test.
+    - Adds a dummy ML model for test purposes.
+    - Tears down the database session and cleans up resources after each test.
+
+    Returns:
+        Generator: A SQLAlchemy session object.
+    """
+    # Create database engine and tables
+    engine = create_engine(TEST_DB_URL)
```
Comment (peterdudfield) [inline tests/conftest.py:36]: I think you should still do  `with PostgresContainer("postgres:15.5") as postgres:` see other examples here https://github.com/openclimatefix/pv-site-datamodel/blob/main/tests/conftest.py#L26

Comment (siddharth7113) [reply tests/conftest.py:36]: Hi @peterdudfield,I’m still looking into this and will make the necessary changes. I’m sorry if it’s taking too long—would it be okay to give me a few more days to finalize it? I’ll update you as soon as it’s done.

**Comment by peterdudfield** [comment]
Thanks for doing those changes, just one more test to fix. If I get time, ill try to look why its failing.

**Comment by siddharth7113** [comment]
> Thanks for doing those changes, just one more test to fix. If I get time, ill try to look why its failing.

No worries! I think I might have an idea why the tests are failing. It could be an issue both with the test I wrote and the code logic itself. I have my university exams in the next few days, so if it's okay with you, I'd like to revisit it afterward and make the necessary corrections.

Once again, thank you for your patience—I’ve learned a lot through this PR process!

**Comment by peterdudfield** [comment]
> > Thanks for doing those changes, just one more test to fix. If I get time, ill try to look why its failing.
> 
> No worries! I think I might have an idea why the tests are failing. It could be an issue both with the test I wrote and the code logic itself. I have my university exams in the next few days, so if it's okay with you, I'd like to revisit it afterward and make the necessary corrections.
> 
> Once again, thank you for your patience—I’ve learned a lot through this PR process!

no problem, good luck and take your time

**Comment by siddharth7113** [comment]
Hi @peterdudfield,

I've been trying to debug the issue, but I still can't figure out what's going wrong. What's particularly interesting is that I'm seeing different test failures locally compared to the CI environment.

One of the key issues seems to be related to getting only one `ForecastSQL` object and populating it with `ForecastValues`. For some reason, this isn't working as expected, and it's creating multiple `ForecastSQL` objects instead.

I was thinking of closing this PR since I've included a bunch of changes together, which might be adding to the confusion. Instead, I'd like to focus solely on the formatting functionality and a test for it first. Would that approach work for you?

Thanks!

**Comment by peterdudfield** [comment]
I think that's a good idea. The save part can cone afterwards On 29 Dec 2024 08:27, Siddharth ***@***.***> wrote:
Hi @peterdudfield,
I've been trying to debug the issue, but I still can't figure out what's going wrong. What's particularly interesting is that I'm seeing different test failures locally compared to the CI environment.
One of the key issues seems to be related to getting only one ForecastSQL object and populating it with ForecastValues. For some reason, this isn't working as expected, and it's creating multiple ForecastSQL objects instead.
I was thinking of closing this PR since I've included a bunch of changes together, which might be adding to the confusion. Instead, I'd like to focus solely on the formatting functionality and a test for it first. Would that approach work for you?
Thanks!

**Pull request closed without merging**